### ISSUE TITLE
Known state ?

### ISSUE THREAD

**Issue by Jeern** [jbogard/Respawn#27]
According to your documentation when I create a Checkpoint I can return to that state when calling the Reset method. However when I read your code in the master branch it seems that you just delete everything in the tables concerned.

If that is the case I suggest you change the documentation.

If this was really about returning to the state at the checkpoint it would be really interesting. I suspect it could be done by Creating a wrapper around the DbConnection, like it is done with the StackExchange MiniProfiler and creating reverse commands of every insert/update and delete statement. It would require "selects" though in order to get the content of the rows before the change. For small databases an alternate approach could be to load the entire database into memory at the checkpoint and dumping it to the tables afterwards.

About your approach with reading Foreign Keys I am not sure a brute force approach is not just as fast or at least only slightly slower, I once did that with a Firebird database. Just deleting everything from all tables. Some would then throw exceptions, I would catch those and delete them again in a second run and so forth.

**Comment by jbogard** [comment]
You're right, I've updated the documentation to say a "clean state".

I've tried various ways of creating an actual checkpoint for known state. They're all horribly slow. Of all the ways I've tested, this manner is the fastest way of deleting. I've tried many, many different ways and they're all slower than deterministic deletion. try-catch works, but I'd need to memoize the order of deletion. try-catch on every test would be slower.

Basically everything is slower than this, other than NOT deleting anything and making your tests resilient to existing data. That's super fast :)

**Comment by Jeern** [comment]
Good to hear you have considered all options to gain max. performance. Are you interested in a pull Request for a Firebird adapter ? I have only skimmed the code but it seems you have chosen a clever path and only use the generic DbConnection, so only the Sql differ.

**Comment by jbogard** [comment]
Yep of course! I might need to switch the CI build to use containers
however.

**Comment by Jeern** [comment]
Ok. But I assume that should not affect the code ? 

> Den 29. nov. 2017 kl. 19.11 skrev Jimmy Bogard <notifications@github.com>:
> 
> Yep of course! I might need to switch the CI build to use containers
> however.
>

**Comment by jbogard** [comment]
No but you’ll need to add tests for Firebase.

**Comment by Jeern** [comment]
Ok. I will see if it is doable. BTW I am not talking about Firebase, but Firebird which we use extensively at my work place. 

> Den 29. nov. 2017 kl. 23.58 skrev Jimmy Bogard <notifications@github.com>:
> 
> No but you’ll need to add tests for Firebase.
>

**Comment by jbogard** [comment]
lol duh, had Firebase on the mind.

If you just get it working with a container (see the readme, should just be
adding a section in the docker compose yml file), I can try to get things
going in AppVeyor.
